### PR TITLE
test: add screenshot test

### PR DESCRIPTION
**Pull request by maciej-ka**
Closes:   https://github.com/styled-components/styled-components/issues/229

Introduces `yarn test:integration` task that will compare example site with saved screenshot. Eventual differences (on test fail) will be visible as png file created in `integration-test/__image_snapshots__/__diff_output__`.

To replace expected image: delete it from `integration-test/__image_snapshots__` and run `yarn test:integration` which will recreate and save png.

### CODE REVIEW

**Comment by maciej-ka** [comment]
Right, I wasn't sure what extras you would need so implementation is minimalistic. Of course, I'll add things you mentioned.

**Comment by maciej-ka** [comment]
Remarks applied

**Comment by mxstbr** [comment]
Uhh, the screenshot is just white?

**Comment by maciej-ka** [comment]
Uuuups, investigating ..., (global animation rule perhaps?)

**Comment by maciej-ka** [comment]
Ach, Example requires external babel links. I removed external request block. We could add those URLs to whitelist and possibly update that whitelist after edits in the example, but I'm thinking: is there a harm in allowing external requests?

**Comment by maciej-ka** [comment]
Added a regexp whitelist

File: integration-test/example.test.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    page.on('request', req => {
+      if (urlWhitelist.find(regexp => req.url.match(regexp))) {
+        req.continue()
+      } else {
+        throw new Error(req.url)
+        req.abort()
+      }
+    })
+    await page.setRequestInterception(true)
+    await page.setViewport({ width: 1024, height: 768 })
```
Comment (morajabi) [inline integration-test/example.test.js]: Mind adding the `deviceScaleFactor` too?
```
await page.setViewport({ width: 1024, height: 768, deviceScaleFactor: 1 })
```

Comment (maciej-ka) [reply integration-test/example.test.js]: Added. Btw. `1` is probably a default option: https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pageemulateoptions

Comment (morajabi) [reply integration-test/example.test.js]: Oh! Good catch :)

**Comment by maciej-ka** [comment]
I'm preparing video training with testing as one chapter and it would be really a value for me to understand, why we put these options in. @philpl, @morajabi Could you tell why not having these could break a test?

I tried to find out in docs, but man, I don't even understand what they do. So pathetic of me.

**--disable-lcd-text**
> Disables LCD text.
> https://peter.sh/experiments/chromium-command-line-switches/#disable-lcd-text

First thought: if you run this test on a cash register then its blue Hitachi LCD Display will have no text. The screenshot will be all blue and test will pass.

**--no-sandbox**
> Disables the sandbox for all process types that are normally sandboxed
> https://peter.sh/experiments/chromium-command-line-switches/#no-sandbox

First thought: it will delete sandbox mode from your Minecraft. You will have more time to improve code quality and tests will pass more often.

**--disable-setuid-sandbox**
> Disable the setuid sandbox (Linux only).
> https://peter.sh/experiments/chromium-command-line-switches/#disable-setuid-sandbox

First thought: it will run GTA without a map. Not sure about impact on running tests.

**deviceScaleFactor**
> <number> Specify device scale factor (can be thought of as dpr). Defaults to 1
> https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#pagesetviewportviewport

This one I almost understand. It's about avoiding image to be blurry on some devices, right? But why value `1` would prevent it? 

Seriously, I'm clueless and would ❤️ to learn something here.

File: integration-test/example.test.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+expect.extend({ toMatchImageSnapshot })
+
+const PORT = 9000
+
+const urlWhitelist = [
+  new RegExp(`http:\/\/localhost:${PORT}.*`),
+  /https:\/\/unpkg\.com\/react@[\d\.]+\/dist\/react\.min\.js/,
+  /https:\/\/unpkg\.com\/react-dom@[\d\.]+\/dist\/react-dom\.min\.js/,
+  /https:\/\/unpkg\.com\/react-dom@[\d\.]+\/dist\/react-dom-server\.min\.js/,
+  /https:\/\/cdnjs\.cloudflare\.com\/ajax\/libs\/babel-core\/[\d\.]+\/browser\.min\.js/
```
Comment (philpl) [inline integration-test/example.test.js]: This is something I'll address in a future change to the dev example, by updating the webpack/server setup. For now this looks like a great workaround :+1:

File: integration-test/example.test.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  let server
+  let browser
+  let page
+
+  beforeAll(async () => {
+    server = await startServer(app, PORT)
+    browser = await launch({
+      executablePath: process.env.GOOGLE_CHROME_BINARY || undefined
+    })
+    browser = await launch({
```
Comment (philpl) [inline integration-test/example.test.js]: Duplicated launch here :wink:

File: integration-test/example.test.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      executablePath: process.env.GOOGLE_CHROME_BINARY || undefined,
+      args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-lcd-text']
```
Comment (philpl) [inline integration-test/example.test.js]: Regarding your comment:

> `'--no-sandbox', '--disable-setuid-sandbox'`

These help with running in a `sudo` Travis environment. I'll make the appropriate changes in a follow-up PR to enable the test on Travis.

> `--disable-lcd-text`

This in combination with a forced `deviceScaleFactor` (can be `> 1` on some Retina/High-Res devices by default) prevents some diverging font rendering. Not perfect but close enough.

Comment (morajabi) [reply integration-test/example.test.js]: Wow it's clear how much you struggled to come up w/ the combination 😄 💯

**Comment by mxstbr** [comment]
> I'll make the appropriate changes in a follow-up PR to enable the test on Travis.

Wait so this isn't running on Travis yet? I is confused

**Comment by maciej-ka** [comment]
Probably we just need to add `script: yarn test:integration` to `.travis.yml` ... and set some env variables.

File: integration-test/example.test.js (hunk 4 above)
Comment (maciej-ka) [reply integration-test/example.test.js]: Thanks!

**Comment by morajabi** [comment]
I think @philpl struggled with Travis a lot for the website. @mxstbr No?

**Comment by philpl** [comment]
@maciej-ka I’ll take care of the travisconfig changes; no worries 😬 it has been tricky to make it work reliably, but fortunately I’ll be able to move the website’s config changes over 1:1

**Comment by mxstbr** [comment]
Thanks a lot for contributing to styled-components! I've invited you to our organization, if you have any questions, let me know!

**Comment by morajabi** [comment]
@maciej-ka I love this moment on Github! 😍  @mxstbr

**Comment by maciej-ka** [comment]
WOW!

**Pull request merged**